core/src/protocols_handler/mod.rs

```diff
@@ -108,22 +108,35 @@ pub trait ProtocolsHandler {
     /// and will be passed back in `inject_substream` or `inject_outbound_closed`.
     type OutboundOpenInfo;
 
-    /// Produces a `ConnectionUpgrade` for the protocol or protocols to accept when listening.
+    /// The timeout for protocol negotiation on inbound substreams.
+    fn inbound_timeout(&self) -> Duration {
+        Duration::from_secs(10)
+    }
+
+    /// The timeout for protocol negotiation on outbound substreams.
+    fn outbound_timeout(&self) -> Duration {
```

Comment:
Similarly, what about adding a `timeout` field to `ProtocolsHandlerEvent::OutboundSubstreamRequest` instead?

Reply:
I can try going down that route, together with the change to `listen_protocol` that you suggested, though it does mean a breaking API change for `ProtocolsHandler`, if we're ok with that.